include/boost/regex/v4/fileiter.hpp

```diff
@@ -229,14 +229,15 @@ class BOOST_REGEX_DECL mapfile_iterator
       file = f;
       node = f->_first + arg_position / mapfile::buf_size;
       offset = arg_position % mapfile::buf_size;
-      file->lock(node);
+      if(file && node)
```

Comment:
On 03/11/2015 20:35, mosherubin wrote:
>
> In include/boost/regex/v4/fileiter.hpp 
> <https://github.com/boostorg/regex/pull/22#discussion_r43803170>:
>
> > @@ -229,14 +229,15 @@ class BOOST_REGEX_DECL mapfile_iterator
> >        file = f;
> >        node = f->_first + arg_position / mapfile::buf_size;
> >        offset = arg_position % mapfile::buf_size;
> > -      file->lock(node);
> > +      if(file && node)
>
> The three functions:
>
>  1. mapfile_iterator(const mapfile* f, long arg_position)
>  2. mapfile_iterator(const mapfile_iterator& i)
>  3. ~mapfile_iterator()
>
> are not consistent in their manner of validity checking. If 
> BOOST_ASSERT is like its C counterpart, then it will only catch a null 
> 'f' value in debug, not release. For release mode, all validity checks 
> must be in the form of C++ comparison statements. For that matter, the 
> BOOST_ASSERT() macro should be used in all three functions.
>
> In addition, the check of "if (file && node)" was being done in only 
> function no. 1, not in 2 and 3. IMHO, the check should be done in all 
> three functions (and not in 1 & 2 only as I submitted).
>
> Both "file" and "node' should be checked for null values in all three 
> functions.
>

You need to realise that these are strictly internal implementation 
details that are only constructed in certain ways (they're also 
deprecated so we're arguing over angels on a pinhead, but never mind).

The following invariants hold:

if file is NULL then node is NULL.
If file is not NULL then node is not NULL.
When constructing from a mapfile then the argument is never NULL.

IMO using asserts to check these invariants is exactly the correct thing 
to do.

Best, John.